### PR TITLE
[Don't forward] LPS-121924 Migrate v2 usages in users-admin/users-admin-web

### PR DESCRIPTION
**Pull request by julien**
The `<clay:management-toolbar-v2 />` tag has been replaced with the
`<clay:management-toolbar />` tag (which uses clay v3).

Creating users and organizations as well a deleting them
and adding users to organizations works as expected.

**Test plan**:
- From the global application menu select the **Control Panel** tab
- Under **Users**, select **Users and Organizations**
- Check that creating a new user by clicking on the "+" button works
- From the management toolbar check that deactivating the user works
  both from the action button or the dropdown menu
- Select the "Organizations" tab
- From the management toolbar check that removing an organization works
  both from the action button or the dropdown menu
- Select an existing organization by clicking on it's name
- From the management toolbar check that adding a nested user works
- From the management toolbar check that adding a nested organization works
- From the management toolbar check that deactivating, removing or restoring nested
  users or organizations works both from the action button or the dropdown menu

**Additionally**
A new callback has been added to the management toolbar's creation menu
to handle creation menu items that are not links

cc @markocikos @ambrinchaudhary @jonmak08 

Please note that since I made changes to the management toolbar's creation menu, I'd like this to be reviewed internally before forwarding.

### CODE REVIEW

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js [hunk 1]
```diff
@@ -190,7 +197,13 @@ const ItemList = ({primaryItems, secondaryItems, visibleItemsCount}) => {
 				}
 
 				return (
-					<ClayDropDown.Item href={item.href} key={index}>
+					<ClayDropDown.Item
+						href={item.href}
```
Comment (ambrinchaudhary) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js:201]: If we want to support links with an `onClick`, I am ok with this. Otherwise, I think it would be better to be restrictive, by setting only the `href` or the `onClick`

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js:201]: That's a very good question @ambrinchaudhary.

I understand your concern and when adding this change I also thought about it.

I think that we could support both `href` and `onClick` since `<ClayDropDown.Item>` (or `<ClayLink>`) supports both, but we can restrict if you think it's better.

Comment (markocikos) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js:201]: @julien We have similar logic in [action controls](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/ActionControls.js#L40-L67):
```
item.href ? <ClayLink ... /> : <ClayButtonWithIcon ... />
```

We should be able to define `onClick` on both link and button. If `href` is not defined, we show a button. This matches [Clay v2](https://github.com/liferay/clay/blob/2.x/packages/clay-dropdown/src/ClayDropdownItem.soy#L140-L165) and should cover all cases in usages.

[Clay v3](https://github.com/liferay/clay/blob/master/packages/clay-drop-down/src/Item.tsx#L66-L67) also supports a `span`, if there is no `href` and `onClick`. There are no usages with this case, but we can pretty painlessly add support for it. We can just copy the logic from Clay v3. 

Can we use `ClayDropDownItem` directly? :thinking: This would simplify things a lot. So instead of 
```
item.href ? <ClayLink ... /> : <ClayButtonWithIcon ... />
```
we would have 
```
<ClayDropDownItem href={} onClick={} ...  />
```

If not, we can extract the logic into something like `DropdownItemElement.js`, and use it both from this file and from `ActionControls.js`

**Review by markocikos — COMMENTED** [review]
Usage update LGTM! The one comment is base tag improvement.

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js (hunk 1 above)
Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js:201]: @markocikos, there are somethings I don't understand

> We should be able to define onClick on both link and button

Yes, that's already the case right?

> Can we use ClayDropDownItem directly? thinking This would simplify things a lot. 

Isn't that the case?

Comment (markocikos) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js:201]: @julien Yeah, you're right, we already have all that. Sorry, I was looking too much in `ActionControls`, not enough in file in this PR :slightly_smiling_face:. `ActionControls` is the one that could be simplified, with similar logic to `ClayDropDown.Item`. But that's outside of scope of this PR:

All good here :+1: 

I would only consider extracting the item into a private component, so we don't have duplication:
```js
const Item = ({...}) => {
	return (<ClayDropDown.Item ... />);
};
```

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js:201]: Hey @markociko no problem, that happens to me too :smile: 

> I would only consider extracting the item into a private component, so we don't have duplication

Sounds like a good idea yes, let me look into that and send another commit (which I'll probably squash before forwarding but I don't want to loose this conversation).

**Review by ambrinchaudhary — APPROVED** [review]

**Comment by jonmak08** [comment]
LGTM!  Thanks @julien!

**Comment by julien** [comment]
Thanks for the review everyone, will forward to @drewbrokke tomorrow

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js [hunk 2]
```diff
@@ -179,6 +179,14 @@ const CreationMenu = ({
...
 };
 
+const Item = ({children, href, onClick}) => {
+	return (
+		<ClayDropDown.Item href={href} onClick={onClick}>
+			{children}
+		</ClayDropDown.Item>
+	);
+};
+
```
Comment (markocikos) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js:189]: @julien We should be able to reduce duplication even further, if we pass `item` prop. If we have

```js
const Item = ({item, key}) => {
	return (
		<ClayDropDown.Item 
                    href={item.href}
                    key={key}
                    onClick={(event) => {
                        onCreationMenuItemClick(event, {item});
                    }}
                >
                        {item.label}
		</ClayDropDown.Item>
	);
};
```
we can invoke it from two places with something like
```js
<Item item={item} key={index} />
```

I might be missing something that does not let us do this :thinking:

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js:189]: You're right that does look even better. 

I see that you're suggesting passing a `key` prop, and in my understanding (which might be wrong) `key` is a special prop (like `ref`) so based on the code from above, what do you you think about a slight change?  

```js
const Item = ({item}) => {
	return (
		<ClayDropDown.Item 
                    href={item.href}
                    onClick={(event) => {
                        onCreationMenuItemClick(event, {item});
                    }}
                >
                        {item.label}
		</ClayDropDown.Item>
	);
};
```

And then in those two places where it's used do

```js
<Item item={item} key={index} />
```

I know it seems strange to pass a key to the `Item` component (without it being used on declared inside that component) but React uses this internally. [This article](https://reactjs.org/warnings/special-props.html) or rather paragraph touches a bit on the topic, but I'd need to read more to actually know all the details.

To go a bit further, I also think that we have to "pass down" the `onCreationItemMenuClick` handler (but correct me if I'm wrong), because it's passed from `ManagementToolbar` > `CreationMenu` > `ItemList` > `Item` , which is going to make the code look like this (at least in my head) 

```javascript
const Item = ({item, onCreationMenuItemClick}) => {
	return (
		<ClayDropDown.Item 
                    href={item.href}
                    onClick={(event) => {
                        onCreationMenuItemClick(event, {item});
                    }}
                >
                        {item.label}
		</ClayDropDown.Item>
	);
};
```

```js
<Item item={item} key={index} onCreationMenuItemClick={onCreationMenuItemClick} />
```

Looks a bit ugly right?

So we could then say: let's pass this as `onClick` and use `...otherProps` in our Item component. 

```js
const Item = ({item, ...otherProps}) => {
	return (
		<ClayDropDown.Item 
                    href={item.href}
                   {...otherProps}
                >
                        {item.label}
		</ClayDropDown.Item>
	);
};
```

And ... 

```js
<Item
     item={item} 
     key={index} 
     onClick={(event) => {
         onCreationMenuItemClick(event, {item});
    }} 
/>
```

It might not be the right time to do yet something else that would avoid passing down these props which is using `useContext` ... but I think we could consider it for the near future, what do you think?

Comment (markocikos) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js:189]: @julien 

You are very much right on `key`, nice catch :+1: 

`onCreationMenuItemClick` will not be available if we put it in file root. But, we can add it as an internal component, like we are doing with [ItemWithLabel here](https://github.com/liferay/liferay-portal/blob/master/modules/apps/app-builder/app-builder-web/src/main/resources/META-INF/resources/js/components/select-objects/SelectObjects.es.js#L111-L133). In fact, we can also do this with `ItemList` and remove its props.

Placing components in root implies they are separate components, only wired through props. This would makes sense if they are imported and used directly, like `ClayDropDown.ItemList` or `ClayDropDown.Item`. But, in our case, `ItemList` and `Item` are just internal helper components, used to squash duplication.

In the same way, `...otherProps` would make sense if `<Item>` was a generic component and we want flexible API. We have it in Clay in most or all components, and it's a good pattern. But, if `<Item>` is an internal component, we control 100% of use cases and there is no need for it.

You mentioned in Slack a different usage case where we don't support item icons and we should. This is an opportunity to also add this:
```js
<ClayDropDown.Item 
     symbolLeft={item.icon}
     ...
>
```

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js:189]: > onCreationMenuItemClick will not be available if we put <Item> in file root. But, we can add <Item> as an internal component, like we are doing with ItemWithLabel here. In fact, we can also do this with ItemList and remove its props.

Hey @markocikos , sorry for not being clear enough.

When you say **in file root**, do you mean putting `Item` in it's own file?

I'm mostly asking because my point was that adding this:

```js
const Item = ({item}) => {
	return (
		<ClayDropDown.Item 
			href={item.href}
			onClick={(event) => {
				onCreationMenuItemClick(event, {item});
			}}
			>
			{item.label}
		</ClayDropDown.Item>
	);
};
```

and then adding it `ItemList`: 

```js
const ItemList = ({
	onCreationMenuItemClick,
	primaryItems,
	secondaryItems,
	visibleItemsCount,
}) => {
	let currentItemCount = 0;

	return (
		<ClayDropDown.ItemList>
			{primaryItems?.map((item, index) => {
				currentItemCount++;

				if (currentItemCount > visibleItemsCount) {
					return false;
				}

				return (
					<Item item={item} key={index} />
				);
			})}

			{secondaryItems?.map((secondaryItemsGroup, index) => (
				<ClayDropDown.Group
					header={secondaryItemsGroup.label}
					key={index}
				>
					{secondaryItemsGroup.items.map((item, index) => {
						currentItemCount++;

						if (currentItemCount > visibleItemsCount) {
							return false;
						}

						return (
							<Item item={item} key={index} />
						);
					})}

					{secondaryItemsGroup.separator && (
						<ClayDropDown.Item className="dropdown-divider" />
					)}
				</ClayDropDown.Group>
			))}
		</ClayDropDown.ItemList>
	);
};
```

In [CreationMenu,js](https://github.com/liferay/liferay-portal/blob/9272e31bc0297f9ad114de2c3a67f68aec8ea3cd/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js) will not "work". It might compile correctly, but at runtime, clicking on the `Item` will give you a runtime error like this



Concerning using other `...otherProps`, I know that this is component we control at 100%, but I think using React's common patterns still makes sense, i.e. Since we have to pass the `onCreationMenuItemClick` callback anyway, I'd rather use this 

```js
const Item = ({item, ...otherProps}) => {
	return (
		<ClayDropDown.Item 
                    href={item.href}
                   {...otherProps}
                >
                        {item.label}
		</ClayDropDown.Item>
	);
};
```

And  

```js
<Item
     item={item} 
     key={index} 
     onClick={(event) => {
         onCreationMenuItemClick(event, {item});
    }} 
/>
```

Rather than 

```js
const Item = ({item, onCreationMenuItemClick}) => {
	return (
		<ClayDropDown.Item 
                    href={item.href}
                   onClick={(event) => {
                       onCreationMenuItemClick(event, {item});
                   }}
                >
                        {item.label}
		</ClayDropDown.Item>
	);
};
```

And 

```js
<Item item={item}  key={index} onCreationMenuItemClick={onCreationMenuItemClick} />
```

But my at the same time I want **everybody** (on the team not on the planet :joy:) to be OK with that, because I think it's important.

Concerning "icon support": good call! I think it's pretty trivial so I can add that too and we'll kill two birds with one stone.

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js]: You can view this [here](https://github.com/liferay-frontend/liferay-portal/pull/774/commits/f0a2180df46063aad30f5fca0447b26f566eb812) or if we prefer 

```js
const Item = ({item, onCreationMenuItemClick}) => {
	return (
		<ClayDropDown.Item 
                    href={item.href}
                   symbolLeft={item.icon}
                   onClick={(event) => {
                       onCreationMenuItemClick(event, {item});
                   }}
                >
                        {item.label}
		</ClayDropDown.Item>
	);
};
```

I can always push another commit on top, and squash before forwarding

Comment (markocikos) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js]: @julien Everything you are suggesting makes sense if we are placing the `Info` component in file root. You are doing this in this PR:
```js
const CreationMenu = ({...}) => {
    return (
        <Item ... /> // we need to add `onCreationMenuItemClick` etc.
    )
}

const Item = ({...}) => {
    return (...)
}
```
Both are in the root of the `CreationMenu.js` file.

What I am suggesting is to make `Item` an internal private component:
```js
const CreationMenu = ({...}) => {
    const Item = ({...}) => {
        return (...)
    }

    return (
        <Item ... /> // we only need `item` prop
    )
}
```
[ItemWithLabel](https://github.com/liferay/liferay-portal/blob/master/modules/apps/app-builder/app-builder-web/src/main/resources/META-INF/resources/js/components/select-objects/SelectObjects.es.js#L111-L133) is an example where we are doing this.

The placement in root vs inside has many implications, as I mentioned in https://github.com/liferay-frontend/liferay-portal/pull/774#discussion_r574402978. `otherProps` makes sense to me in former, but not the latter. I think that, for our purpose, it makes more sense to place both `Item` and `ItemList` inside.

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js]: @markocikos OK got it, now it all makes more sense for me too :+1: 

In this case and as you know I don't like the fact that the `ItemList` and `Item` function are going to be re-declared each time `CreationMenu` is going to be invoked, but OK, I'll do that.

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js]: @markocikos see [
fd50016](https://github.com/liferay-frontend/liferay-portal/pull/774/commits/fd5001602931bab69656dab310a6fc0d8809faa3) and tell me if that's OK when you can - thanks. I'll squash before forwarding.

Comment (markocikos) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js]: I'm not sure, maybe there is an issue here :thinking: 

Summoning @wincent ! Is it safe to use React components within components? It makes sense from separation of concerns perspective, but is there a performance or a React quirk issue here?

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js]: @markocikos just so you know, @wincent will be on PTO tomorrow, so if this is not a big risk (and I think it isn't) we could probably pass this as is (if you agree) and discuss this later on ... if it needs to be changed then we can do it in another PR - What are your thoughts on that? Thanks.

Comment (markocikos) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js]: @julien Yeah I agree. A performance issue is not urgent, if there is an issue with it we can iterate later on. And if there is a quirk with Reach that prevents this from working, we would see it already.

Let's push this! :rocket:

**Review by markocikos — APPROVED** [review]

**Comment by julien** [comment]
Thanks for the update @markocikos. I'll rebase, squash and forward.

**Comment by julien** [comment]
Forwarded [here](https://github.com/drewbrokke/liferay-portal/pull/1301)

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js (hunk 2 above)
Comment (wincent) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js]: > Summoning @wincent ! Is it safe to use React components within components? It makes sense from separation of concerns perspective, but is there a performance or a React quirk issue here?

tl;dr: Rule of thumb — prefer to hoist components up to the top-level when you can (for performance, to reduce GC pressure, to make clearer what the responsibility of each component is) _but_ as soon as there is a reason to inline the component (eg. because it'll be way more convenient to access state via a closure instead of via explicit prop passing) it is totally fine to inline.

---

The bug-or-feature here is that because React components are just functions the same considerations apply as for _any_ function that is defined inside another function. It is going to close over the variables in its scope, which means you can access them without having to pass them explicitly. On the plus side this might make it easier to use the components. On the negative side, you might wind up closing over "stale" state which could lead to a bug. Whether the closure helps or hurts you depends on the context.

And pretty much exact same deal for performance. By putting the function inside the other component it is getting recreated on every render. It will be short-lived and get garbage-collected, which could be wasteful. By hoisting it outside to the module root you have exactly one copy of it that gets reused for the lifetime of the program. Does this matter in practice? The answer is "probably not", but if you can hoist the function up without sacrificing too much (ie. because you don't really need to pass things via closure) then you should.

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js]: Thanks for the advice @wincent
